File: macos/SeedSeeker/Tests/SeedSeekerKitTests/EngineConstantsTests.swift
import CSeedFinder
import Foundation
import XCTest
@testable import SeedSeekerKit

/// The app keeps local copies of the engine's scalar constants so the models
/// need nothing from the engine to validate. This is the one place they meet
/// the engine: every local is asserted against the `seedfinder_engine_info`
/// document the linked engine publishes, so a change on either side fails
/// here rather than as an editor offering a query the search refuses.
final class EngineConstantsTests: XCTestCase {
    /// Read per test case: the document is a constant of the linked engine
    /// and cheap to fetch, and an instance property keeps the test free of
    /// shared mutable state under strict concurrency.
    private let info: [String: Any] = {
        let packet = try! enginePacket { out, length in seedfinder_engine_info(out, length) }
        return try! JSONSerialization.jsonObject(with: packet) as! [String: Any]
    }()
    private var limits: [String: Any] { info["limits"] as! [String: Any] }
    private func limit(_ key: String) -> Int { limits[key] as! Int }

    func testQueryBoundsMatchTheEngine() {
        XCTAssertEqual(SearchLimits.maxDepth, limit("maxDepth"))
        XCTAssertEqual(SearchLimits.exactTiers, limit("exactTierMin")...limit("exactTierMax"))
        XCTAssertEqual(SearchLimits.boundedTiers, limit("boundedTierMin")...limit("boundedTierMax"))
        XCTAssertEqual(SearchLimits.identityGroupMax, limit("identityGroupMax"))
        XCTAssertEqual(SearchLimits.levelSumGroupMax, limit("levelSumGroupMax"))
        XCTAssertEqual(SearchLimits.maxUpgradeDefault, limit("maxUpgradeDefault"))
        XCTAssertEqual(SearchLimits.maxUpgradeRing, limit("maxUpgradeRing"))
        XCTAssertEqual(SearchLimits.maxUpgradeRingStandard, limit("maxUpgradeRingStandard"))
        XCTAssertEqual(SearchLimits.maxUpgradeWeapon, limit("maxUpgradeWeapon"))
        // The families route to the right maximum, narrowed weapon kinds
        // included: the engine publishes one ceiling per family.
        let byKind = limits["maxUpgradeByKind"] as! [String: Int]
        let names: [ItemKind: String] = [.weapon: "weapon", .armor: "armor", .wand: "wand",
                                         .ring: "ring", .meleeWeapon: "weapon",
                                         .thrownWeapon: "weapon", .trinket: "trinket", .artifact: "artifact"]
        for kind in ItemKind.allCases {
            XCTAssertEqual(kind.maximumSearchUpgrade, byKind[names[kind]!], "\(kind)")
        }
        XCTAssertEqual(SearchLimits.maxUpgradeAnyTier, limit("maxUpgradeAnyTier"))
        XCTAssertEqual(SearchLimits.extraUpgradeTier, limit("extraUpgradeTier"))
    }

    /// Only a tier-4 weapon is levelled past `maxUpgradeAnyTier`, so a
    /// requirement that rules that tier out loses the top of its range.
    @MainActor
    func testTopWeaponUpgradeNeedsTheTierThatReachesIt() {
        let extraTier = limit("extraUpgradeTier")
        let ceiling = limit("maxUpgradeWeapon")
        let capped = limit("maxUpgradeAnyTier")
        let maximum = { (item: CatalogItem?, tier: Int, match: TierMatch) in
            SearchLimits.maximumUpgrade(kind: .weapon, item: item, tier: tier, tierMatch: match)
        }
        XCTAssertEqual(maximum(nil, 0, .any), ceiling)
        XCTAssertEqual(maximum(nil, extraTier, .exactly), ceiling)
        XCTAssertEqual(maximum(nil, 5, .exactly), capped)
        XCTAssertEqual(maximum(nil, 3, .atMost), capped)
        XCTAssertEqual(maximum(ItemCatalog.findById("battle_axe"), 0, .any), ceiling)
        XCTAssertEqual(maximum(ItemCatalog.findById("javelin"), 0, .any), ceiling)
        XCTAssertEqual(maximum(ItemCatalog.findById("sword"), 0, .any), capped)
        XCTAssertEqual(SearchLimits.maximumUpgrade(kind: .armor, item: nil, tier: extraTier, tierMatch: .exactly), capped)
    }

    @MainActor
    func testSessionAndFileLimitsMatchTheEngine() {
        XCTAssertEqual(SearchController.resultCap, info["maxResults"] as? Int)
        // The import byte cap and the seed count have no local copies: the
        // codec applies the cap itself and the session reports the count.
        XCTAssertGreaterThan(limit("resultsFileMaxBytes"), 0)
        XCTAssertEqual(info["totalSeeds"] as? Int64, 5_429_503_678_976)
        XCTAssertEqual(EngineInfo.shared.shpdVersion, info["shpdVersion"] as? String)
    }

    func testEmptyBossFloorsMatchTheEngine() {
        XCTAssertEqual(FloorLimits.emptyBossFloors, Set(info["emptyBossFloors"] as! [Int]))
        XCTAssertEqual(FloorLimits.options,
                       (1...SearchLimits.maxDepth).filter { !FloorLimits.emptyBossFloors.contains($0) })
    }

    func testQuestWindowsMatchTheEngine() {
        let windows = info["questWindows"] as! [String: [Int]]
        let names: [ScoutQuestKind: String] = [.ghost: "ghost", .wandmaker: "wandmaker",
                                               .blacksmith: "blacksmith", .imp: "imp"]
        XCTAssertEqual(Set(windows.keys), Set(names.values))
        for kind in ScoutQuestKind.allCases {
            let window = windows[names[kind]!]!
            XCTAssertEqual(window.count, 2)
            XCTAssertEqual(kind.depthRange, window[0]...window[1], kind.giverLabel)
        }
    }

    func testChallengesMatchTheEngineInMaskOrder() {
        let engine = (info["challenges"] as! [[String: Any]]).map { entry in
            (entry["name"] as! String, entry["mask"] as! Int, entry["changesLevelGeneration"] as! Bool)
        }
        let localNames = ["on_diet", "faith_is_my_armor", "pharmacophobia", "barren_land", "swarm_intelligence",
                          "into_darkness", "forbidden_runes", "hostile_champions", "badder_bosses"]
        let local = zip(localNames, Challenge.allCases).map { ($0, $1.rawValue, $1.changesLevelGeneration) }
        XCTAssertEqual(engine.count, local.count)
        for (index, (engineEntry, localEntry)) in zip(engine, local).enumerated() {
            XCTAssertEqual(engineEntry.0, localEntry.0)
            XCTAssertEqual(engineEntry.1, localEntry.1)
            XCTAssertEqual(engineEntry.2, localEntry.2, engineEntry.0)
            XCTAssertEqual(localEntry.1, 1 << index)
        }
        XCTAssertEqual(SearchLimits.challengeMask, engine.reduce(0) { $0 | $1.1 })
    }
}

File: windows/SeedSeeker.Tests/EngineConstantsTests.cs
using System.Text.Json.Nodes;
using Xunit;

namespace SeedSeeker.Tests;

/// <summary>
/// The app keeps local copies of the engine's scalar constants so the editor
/// needs nothing from the engine to open. This is the one place they meet the
/// engine: every local is asserted against the <c>engine_info</c> document
/// the linked engine publishes, so a change on either side fails here rather
/// than as an editor offering a query the search refuses.
/// </summary>
public sealed class EngineConstantsTests
{
    private static readonly JsonObject Info = (JsonObject)JsonNode.Parse(NativeEngine.EngineInfoJson())!;
    private static readonly JsonObject Limits = (JsonObject)Info["limits"]!;

    private static int Limit(string key) => (int)Limits[key]!;

    [Fact]
    public void QueryBoundsMatchTheEngine()
    {
        Assert.Equal(SearchLimits.MaxDepth, Limit("maxDepth"));
        Assert.Equal(SearchLimits.ExactTierMin, Limit("exactTierMin"));
        Assert.Equal(SearchLimits.ExactTierMax, Limit("exactTierMax"));
        Assert.Equal(SearchLimits.BoundedTierMin, Limit("boundedTierMin"));
        Assert.Equal(SearchLimits.BoundedTierMax, Limit("boundedTierMax"));
        Assert.Equal(SearchLimits.IdentityGroupMax, Limit("identityGroupMax"));
        Assert.Equal(SearchLimits.LevelSumGroupMax, Limit("levelSumGroupMax"));
        Assert.Equal(SearchLimits.MaxUpgradeDefault, Limit("maxUpgradeDefault"));
        Assert.Equal(SearchLimits.MaxUpgradeRing, Limit("maxUpgradeRing"));
        Assert.Equal(SearchLimits.MaxUpgradeRingStandard, Limit("maxUpgradeRingStandard"));
        Assert.Equal(SearchLimits.MaxUpgradeWeapon, Limit("maxUpgradeWeapon"));
        // The families route to the right maximum, narrowed weapon kinds included.
        Assert.Equal(Limit("maxUpgradeRing"), ItemKind.Ring.MaximumSearchUpgrade());
        foreach (var kind in new[] { ItemKind.Weapon, ItemKind.MeleeWeapon, ItemKind.ThrownWeapon })
            Assert.Equal(Limit("maxUpgradeWeapon"), kind.MaximumSearchUpgrade());
        foreach (var kind in new[] { ItemKind.Armor, ItemKind.Wand })
            Assert.Equal(Limit("maxUpgradeDefault"), kind.MaximumSearchUpgrade());
        // Every family's ceiling is the engine's own, keyed by the document's name for it.
        foreach (var (name, kind) in new[] {
            ("weapon", ItemKind.Weapon), ("armor", ItemKind.Armor), ("wand", ItemKind.Wand), ("ring", ItemKind.Ring), ("artifact", ItemKind.Artifact) })
            Assert.Equal((int)((JsonObject)Limits["maxUpgradeByKind"]!)[name]!, kind.MaximumSearchUpgrade());
        Assert.Equal(SearchLimits.MaxUpgradeAnyTier, Limit("maxUpgradeAnyTier"));
        Assert.Equal(SearchLimits.ExtraUpgradeTier, Limit("extraUpgradeTier"));
        Assert.Equal(SearchLimits.MaxDepth, new QuerySettings().MaximumDepth);
    }

    /// <summary>
    /// Only a tier-4 weapon is levelled past the shared ceiling, so a
    /// requirement that rules that tier out loses the top of its range.
    /// </summary>
    [Fact]
    public void TopWeaponUpgradeNeedsTheTierThatReachesIt()
    {
        var ceiling = Limit("maxUpgradeWeapon");
        var capped = Limit("maxUpgradeAnyTier");
        var extraTier = Limit("extraUpgradeTier");
        Assert.Equal(ceiling, ItemKind.Weapon.MaximumSearchUpgrade(null, TierMatch.Any, 0));
        Assert.Equal(ceiling, ItemKind.Weapon.MaximumSearchUpgrade(null, TierMatch.Exactly, extraTier));
        Assert.Equal(capped, ItemKind.Weapon.MaximumSearchUpgrade(null, TierMatch.Exactly, 5));
        Assert.Equal(capped, ItemKind.Weapon.MaximumSearchUpgrade(null, TierMatch.AtMost, 3));
        Assert.Equal(ceiling, ItemKind.Weapon.MaximumSearchUpgrade(ItemCatalog.Find("battle_axe"), TierMatch.Any, 0));
        Assert.Equal(ceiling, ItemKind.ThrownWeapon.MaximumSearchUpgrade(ItemCatalog.Find("javelin"), TierMatch.Any, 0));
        Assert.Equal(capped, ItemKind.Weapon.MaximumSearchUpgrade(ItemCatalog.Find("sword"), TierMatch.Any, 0));
        Assert.Equal(capped, ItemKind.Armor.MaximumSearchUpgrade(null, TierMatch.Exactly, extraTier));
    }

    [Fact]
    public void SessionAndFileLimitsMatchTheEngine()
    {
        Assert.Equal(SearchLimits.ResultCap, (int)Info["maxResults"]!);
        // The import byte cap has no local copy: the app reads it from the
        // engine at runtime (EngineInfo.ResultsFileMaxBytes) and the codec
        // applies it itself. Pin that the runtime reader agrees with the
        // document this test reads.
        Assert.Equal(Limit("resultsFileMaxBytes"), EngineInfo.ResultsFileMaxBytes);
        Assert.Equal((string?)Info["shpdVersion"], EngineInfo.ShpdVersion);
        Assert.Equal((string?)Info["shpdCommit"], EngineInfo.ShpdCommit);
    }

    [Fact]
    public void EmptyBossFloorsMatchTheEngine()
    {
        var floors = ((JsonArray)Info["emptyBossFloors"]!).Select(floor => (int)floor!).ToArray();
        Assert.Equal(floors, FloorLimits.EmptyBossFloors);
        Assert.Equal(Enumerable.Range(1, SearchLimits.MaxDepth).Except(floors), FloorLimits.Options);
    }

    [Theory]
    [InlineData("ghost", QuestGiver.Ghost)]
    [InlineData("wandmaker", QuestGiver.Wandmaker)]
    [InlineData("blacksmith", QuestGiver.Blacksmith)]
    [InlineData("imp", QuestGiver.Imp)]
    public void QuestWindowsMatchTheEngine(string name, QuestGiver giver)
    {
        var window = ((JsonArray)Info["questWindows"]![name]!).Select(floor => (int)floor!).ToArray();
        Assert.Equal(2, window.Length);
        Assert.Equal((window[0], window[1]), ScoutQuests.Window(giver));
    }

    [Fact]
    public void ChallengesMatchTheEngineInMaskOrder()
    {
        var engine = ((JsonArray)Info["challenges"]!)
            .Select(entry => ((string)entry!["name"]!, (int)entry["mask"]!, (bool)entry["changesLevelGeneration"]!))
            .ToArray();
        var local = Challenges.All.Select(entry => (entry.Name, entry.Mask, entry.ChangesLevelGeneration)).ToArray();
        Assert.Equal(engine, local);
        for (var index = 0; index < local.Length; index++) Assert.Equal(1 << index, local[index].Mask);
        Assert.Equal(engine.Aggregate(0, (mask, entry) => mask | entry.Item2), Challenges.AllMask);
    }
}
